src/Ouroboros/Storage/Immutable/DB.hs

```diff
@@ -282,45 +296,192 @@ modifyInternalState db newEpoch action = do
+  current <- atomically $ readTVar _it_current
+  case outOfBounds current _it_end of
+      True  -> runExceptT $ do
+          ExceptT $ iteratorClose it
+          return Nothing
+      False -> do
+          -- Read more data and step the iterator
+          (eHnd, fullIndex) <- atomically $ do
+              e <- readTVar _it_epoch_handle
+              i <- readTVar _it_epoch_index
+              return (e,i)
+          r <- readNext eHnd fullIndex (snd current)
+          runExceptT $ case r of
+              Left err -> liftFsError $ do
+                  hClose eHnd
+                  throwError err
+              Right bs -> do
+                  -- Step the iterator
+                  iteratorStep
+                  return  $ Just (fst current, snd current, bs)
+  where
+    outOfBounds :: (Epoch, RelativeSlot) -> (Epoch, RelativeSlot) -> Bool
+    outOfBounds (currentEpoch, currentSlot) (endEpoch, endSlot)
+      | currentEpoch > endEpoch                           = True
+      | currentEpoch == endEpoch && currentSlot > endSlot = True
+      | otherwise                                         = False
+
+    readNext :: (Monad m, HasFSE m)
+             => FsHandleE m
+             -> U.Vector (Word64, Word64)
+             -> RelativeSlot
+             -> m (Either FsError ByteString)
+    readNext eHnd fullIndex (RelativeSlot slot) = runExceptT $ do
+          -- Step 1: grab from the cached index the blobSize.
+          let (_, !blobSize) = fullIndex U.! (fromIntegral slot)
+
+          -- Step 2: Read from the epoch file. No need for seeking: as we
+          -- are streaming, we are already positioned at the correct place.
+          hGet eHnd (fromEnum blobSize)
+
+    -- Steps the internal state of an 'Iterator'. Internal use only.
+    iteratorStep :: (MonadSTM m, MonadMask m, HasFSE m)
+                 => ExceptT ImmutableDBError m ()
+    iteratorStep = liftFsError $ do
+        ((epoch, RelativeSlot slot), fullIndex) <- atomically $ do
+            (,) <$> readTVar _it_current <*> readTVar _it_epoch_index
+
+        let epoch' = epoch + 1
+        -- If the current relative slot is equal to the size of the index, it
+        -- means we reached the end of the epoch and we need to open the next
+        -- epoch file.
+        case fromIntegral (slot + 1) == U.length fullIndex &&
+             not (outOfBounds (epoch', 0) _it_end) of
+             True  -> do
+                 eHnd <- atomically $ readTVar _it_epoch_handle
+                 eHnd' <- hOpen (_dbFolder db <> renderFile "epoch" epoch') ReadMode
+                          `finally` hClose eHnd
+                 index' <- loadIndex db epoch'
+                 atomically $ do
+                     writeTVar _it_epoch_handle eHnd'
+                     writeTVar _it_epoch_index index'
+                     writeTVar _it_current (epoch', RelativeSlot 0)
+             False -> atomically $ writeTVar _it_current (epoch, RelativeSlot $ slot + 1)
```

Comment:
This is not correct. It should skip non-present blocks.

Reply:
We should have a test property that whatever I write I get out, and nothing less, and notihng more.

Reply:
The same then needs to happen when we create the iterator (to satisfy the invariant).